nixos/syncthing: add missing `--fail` to `curl`

dpc opened the pull request:
This is to address a problem I've found when debugging: https://github.com/NixOS/nixpkgs/issues/326704#issuecomment-2730590839

Without `--fail`, curl does not seem to set exit code, and I'm under suspicion all the other retry flags etc. are doing nothing as well.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dpc committed:
nixos/syncthing: missing `--fail` to `curl`

bjornfor reviewed (commented):
LGTM, but please update commit message to have "nixos/syncthing:" prefix.

dpc committed:
nixos/syncthing: make it easier to debug incorrect IDs

In case of an incorrect ID, the behavior of this script
is quite tricky. See:

https://github.com/NixOS/nixpkgs/issues/326704#issuecomment-2730819822

for more info.

dpc commented:
@bjornfor

I've added one more change to help with debugging. If something's wrong with please let me know, I can always remove.

74k1 commented:
Hey @dpc 

Wanted to chime in and say, that a rebuild would fail because of this when having the following options set to false:

```Nix
services.syncthing.settings = {
  overrideDevice = false;
  overrideFolder = false;
}
```

I get an error for the `syncthing-init.service`.

<details>
<summary>Console Output of manually running the service command</summary>

```
env RUNTIME_DIRECTORY=/tmp bash -x /nix/store/qngqpx8bmwh98hdmkmxipqjyyl6l3lqn-merge-syncthing-config
+ set -efu
+ umask 0077
+ curl -X PUT -d true 127.0.0.1:8384/rest/config/relaysEnabled
+ /nix/store/ag4wrj6qc542q7l67ld6i1pwbcp1cjq4-libxml2-2.13.6-bin/bin/xmllint --xpath 'string(configuration/gui/apikey)' /home/taki/.config/syncthing/config.xml
+ printf 'X-API-Key: '
+ cat /tmp/api_key
+ /nix/store/ipyzblpplw889rdvmm7sw8r3kf24p6c2-curl-8.12.1-bin/bin/curl --fail -sSLk -H @/tmp/headers --retry 1000 --retry-delay 1 --retry-all-errors -X PUT -d true 127.0.0.1:8384/rest/config/relaysEnabled
curl: (22) The requested URL returned error: 404
curl: (22) The requested URL returned error: 404
curl: (22) The requested URL returned error: 404
```

</details>

I'm not _entirely_ sure if it's really related to my syncthing settings, as I wouldn't want my current devices and folders to get removed by testing it on `true`.

At least removing the `--fail` flag on that `curl` worked for me.

It could be that it's just me. If that is the case, please let me know.

bjornfor commented:
I'm seeing the same in my local integration tests (repeated `curl: (22) The requested URL returned error: 404`, and eventually timeout). The NixOS tests don't have/see a problem though. I think we should revert until this is better understodd.

bjornfor commented:
Revert here: https://github.com/NixOS/nixpkgs/pull/392178. Please review.
